improve-ephemeral-cluster-handling.md

```diff
@@ -0,0 +1,419 @@
+      bob-1                1         1         1         8m
+      cluster-ckdtw2-m-1   0         0         0         2s
+      cluster-ckdtw2-w-1   0         0         0         2s
+
+      # Partial export of the deploymentconfig for the spark master
+      $ oc export dc cluster-4a6xhf-m
+      apiVersion: v1
+      kind: DeploymentConfig
+      metadata:
+        creationTimestamp: null
+        generation: 1
+        labels:
+          ephemeral: bob-1
+          oshinko-cluster: cluster-4a6xhf
+          oshinko-type: master
+        name: cluster-4a6xhf-m
+
+* `uses-oshinko-cluster`
+
+  The value of this label is the name of an oshinko cluster. This label
+  goes on the replicationcontroller for the spark driver pod that creates
+  the ephemeral cluster. For example:
+
+      $ oc export rc bob-1
+      apiVersion: v1
+      kind: ReplicationController
+      metadata:
+        creationTimestamp: null
+        generation: 1
+        labels:
+          app: oshinko-pyspark-dc
+          application: oshinko-pyspark
+          createdBy: template-oshinko-pyspark-dc
+          openshift.io/deployment-config.name: bob
+          uses-oshinko-cluster: cluster-ckdtw2
+        name: bob-1
+
+### Adding kubernetes downward api volume mounts
+
+The spark driver deploymentconfigs will use features in the kubernetes downward
+api to mount the label set from the dc on the driver pods as a volume.  This will
+allow the `start.sh` script to look up relevant information stored in labels,
+such as the name of the driver pod's deployment (replicationcontroller)
+
+Knowing the deployment of the driver pod is essential for `start.sh` to manage
+ephemeral clusters correctly. It allows the script to record a generated
+cluster name for the deployment, create an ephemeral cluster associated with
+the deployment, and guarantee that it only deletes an ephemeral cluster
+associated with the deployment. This is accomplished by passing the name of
+the deployment as the value of certain flags in oshinko CLI commands (more below).
+
+### Extensions to oshinko CLI to enable ephemeral cluster features
+
+To make ephemeral cluster handling work from within `start.sh` running
+in a Spark driver pod, we will add options to the `create`, `delete`,
+and `get` oshinko CLI commands:
+
+* Options on create to mark the cluster ephemeral and associate the driver
+
+      $ oshinko-cli create fred --app=<name of driver deployment> --ephemeral
+
+  The value of the `--app` option may be either the name of the driver pod, or the
+  name of the driver's deployment (both are available to `start.sh`). The `--ephemeral`
+  option causes oshinko to mark the cluster as ephemeral (by default clusters are not
+  ephemeral). The values of these options are used by oshinko to set the labels discussed
+  above.
+
+  Note that if the `--ephemeral` option is set, the `--app` option must also be set.
+  However, the `--app` option may be used by itself to associate a driver with a
```

Comment:
This should say "theoretically". At the moment there is no command to associate with a long-running cluster other than at create time. Discussed below, maybe remove this here.